### PR TITLE
Added zoom animation as an option.

### PR DESCRIPTION
**Pull request by manthey**
Added zoom animation as an option. The speed and easing can be adjusted.

When a second transition is queued, we have a one animation frame delay when neither the current nor the new animation was being applied.  This delay has been removed.

Added some code to ensure pixel-perfect alignment in certain instances where it seemed like it didn't happen.

This resolves issue #542.

### CODE REVIEW

File: src/core/map.js [hunk 1]
```diff
@@ -447,7 +458,7 @@ geo.map = function (arg) {
     m_rotation = rotation;
 
     var bounds = m_this.boundsFromZoomAndCenter(
-        m_zoom, m_center, m_rotation, null);
+        m_zoom, m_center, m_rotation, null, ignoreRotationFunc);
```
Comment (aashish24) [inline src/core/map.js]: why we need to pass ignoreRotationFunc now?

**Comment by aashish24** [comment]
@manthey mostly looks good to me. Can you update this branch with current master?

**Comment by aashish24** [comment]
@manthey I believe https://github.com/OpenGeoscience/geojs/pull/549 is merged into this branch? If that's right, should we close #549?

**Comment by manthey** [comment]
@aashish24 I was trying to keep conceptual ideas in different PRs.  #549 could be reviewed and approved first, then this PR.

File: src/core/map.js (hunk 1 above)
Comment (manthey) [inline src/core/map.js]: During a transition, if we in discreteZoom mode and are animating both zoom and rotation, if we don't pass this flag we may end up discretizing the zoom level during the transition.

**Comment by manthey** [comment]
@aashish24 I rebased on master.

File: src/core/map.js [hunk 2]
```diff
@@ -1194,11 +1226,14 @@ geo.map = function (arg) {
    * Cancel any existing transition.  The transition will send a cancel event
    * at the next animation frame, but no further activity occurs.
    *
+   * @param {string} source option cause of the cancel.
```
Comment (jbeezley) [inline src/core/map.js]: Could you describe this argument a bit more.  It seems you are expecting a string like `<method name>.<action>`.  Can this method be called without a `source` argument?

Comment (manthey) [inline src/core/map.js]: Yes, `source` is optional  The description should have read 'optional' (not option).  It falls back to the a previous value (if it was set), and that falls back to an empty string.

I've used it as `<method>.<action>`, but that isn't strictly required.  I'll add to the comment that it is preferred.

Comment (jbeezley) [inline src/core/map.js]: Sounds good.  The [jsdoc syntax](http://usejsdoc.org/tags-param.html) for optional arguments is typically `@param {string} [source]`.

**Comment by jbeezley** [comment]
Testing this out it works well with an ordinary mouse wheel, but it is still feels awkward with touch scrolling to me.  My [Magic Mouse](http://www.apple.com/shop/product/MLA02LL/A/magic-mouse-2) tries to simulate scrolling animation by firing many small scroll events.  These events are internally smoothed out over a period ~1 second which gives the same animated feel that we are emulating in our source.  Using this with my mouse gives the distinct feeling that the mouse is fighting with map's internal animation.

I had attempted to throttle the wheel events to prevent some of this, but I don't think it is sufficient.  For one, the fact that the wheel events continue to fire for up to 1 second after your finger stops scrolling tends to make you overshoot.  In mapbox, there is a [block of code](https://github.com/mapbox/mapbox-gl-js/blob/master/js/ui/handler/scroll_zoom.js#L82-L112) that does more intelligent throttling depending on whether it detects a mousewheel or trackpad that we may have to emulate to unify the experience between devices.

There is also a race condition that I can pretty easily trigger with my mouse that causes the map animation to stop between zoom levels even when discrete zoom is enabled.  It also seems to occasionally remember a queued transition after the animation is stopped.  The next wheel event will start the queued transition before starting the new one.  This means that the zoom will sometimes start animating in the wrong direction briefly before responding to the intended interaction.

**Comment by aashish24** [comment]
> Testing this out it works well with an ordinary mouse wheel, but it is still feels awkward with touch scrolling to me. My Magic Mouse tries to simulate scrolling animation by firing many small scroll events. These events are internally smoothed out over a period ~1 second which gives the same animated feel that we are emulating in our source. Using this with my mouse gives the distinct feeling that the mouse is fighting with map's internal animation.

@manthey I am assuming that you are planning to address these? Please let me know.

**Comment by manthey** [comment]
Yes.  I need to get a Magic Mouse or borrow a Mac laptop to see what events
are being fired and why it behaves oddly there.

On Fri, Mar 25, 2016 at 11:56 AM, Aashish Chaudhary <
notifications@github.com> wrote:

> Testing this out it works well with an ordinary mouse wheel, but it is
> still feels awkward with touch scrolling to me. My Magic Mouse tries to
> simulate scrolling animation by firing many small scroll events. These
> events are internally smoothed out over a period ~1 second which gives the
> same animated feel that we are emulating in our source. Using this with my
> mouse gives the distinct feeling that the mouse is fighting with map's
> internal animation.
>
> @manthey <https://github.com/manthey> I am assuming that you are planning
> to address these? Please let me know.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/OpenGeoscience/geojs/pull/550#issuecomment-201342298>
>



-- 
David Manthey
R&D Engineer
Kitware Inc.
(518) 881-4439

**Comment by manthey** [comment]
@jbeezley The Mac gives zero-distance wheel events, which I wasn't expecting.  Let me know if there are still issues.

**Comment by jbeezley** [comment]
Great, looks like that fixed it.  LGTM.

**Comment by aashish24** [comment]
Yay :+1:

**Pull request merged**